### PR TITLE
soroban-rpc: refactor libpreflight

### PR DESCRIPTION
**Pull request by 2opremio**
* Split out FFI conversions from actual preflight computations. I added a new file `preflight.rs` which contains the actual preflight computations, separate from the `Go<->Rust` ffi interfacing (which is the only thing left in `lib.rs`).
* Use the [`anyhow`](https://docs.rs/anyhow/latest/anyhow/) crate to:
   * Make error handling more readable (`Result<R, Box<dyn Error>>` vs `Result<R>`)
   * Add context to errors ( Closes https://github.com/stellar/soroban-tools/issues/858).
* Some other minor changes, making the code more idiomatic/readable.
* Fix a couple of memory leaks in `free_preflight_result()`

### CODE REVIEW

File: cmd/soroban-rpc/lib/preflight/src/fees.rs [hunk 1]
```diff
@@ -138,13 +141,14 @@ fn calculate_host_function_soroban_resources(
...
+        calculate_event_size_bytes(events).context("cannot calculate events size")?;
 
     // Add a 15% leeway with a minimum of 50k instructions
-    let instructions = max(
-        budget.get_cpu_insns_consumed()? + 50000,
-        budget.get_cpu_insns_consumed()? * 115 / 100,
-    );
+    let budget_instructions = budget
+        .get_cpu_insns_consumed()
+        .context("cannot get instructions consumed")?;
```
Comment (paulbellamy) [inline cmd/soroban-rpc/lib/preflight/src/fees.rs:150]:
```suggestion
        .context("cannot calculate instructions consumed")?;
```
Nit: Might be more accurate? Because, "cannot get" could mean the actual instructions are not available to be consumed. Probably fine either way tho.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs [hunk 2]
```diff
@@ -157,54 +161,43 @@ fn calculate_host_function_soroban_resources(
...
     };
 
     let ConfigSettingEntry::ContractLedgerCostV0(ledger_cost) =
         ledger_storage.get_configuration_setting(ConfigSettingId::ContractLedgerCostV0)?
     else {
-        return Err(
-            "get_fee_configuration(): unexpected config setting entry for LedgerCostV0 key".into(),
-        );
+        bail!(
+           "unexpected config setting entry for LedgerCostV0 key");
```
Comment (paulbellamy) [inline cmd/soroban-rpc/lib/preflight/src/fees.rs:176]: Nit: Is this actually rustfmt? Looks weird.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs [hunk 3]
```diff
@@ -239,20 +232,28 @@ fn calculate_unmodified_ledger_entry_bytes(
...
 }
 
-fn calculate_event_size_bytes(events: &Vec<DiagnosticEvent>) -> Result<u32, Box<dyn error::Error>> {
+fn calculate_event_size_bytes(events: &Vec<DiagnosticEvent>) -> Result<u32> {
     let mut res: u32 = 0;
     for e in events {
-        let event_xdr = e.to_xdr()?;
+        let event_xdr = e
+            .to_xdr()
+            .with_context(|| format!("cannot marshall event {e:?}"))?;
```
Comment (paulbellamy) [inline cmd/soroban-rpc/lib/preflight/src/fees.rs:256]: Do we use `marshal` or `marshall`? I can never remember which we use in this codebase...

File: cmd/soroban-rpc/lib/preflight/src/lib.rs [hunk 4]
```diff
@@ -42,20 +36,22 @@ pub struct CLedgerInfo {
...
 
-impl From<CLedgerInfo> for LedgerInfo {
-    fn from(c: CLedgerInfo) -> Self {
-        let network_passphrase_cstr = unsafe { CStr::from_ptr(c.network_passphrase) };
-        Self {
+impl TryFrom<CLedgerInfo> for LedgerInfo {
+    type Error = anyhow::Error;
+
+    fn try_from(c: CLedgerInfo) -> Result<Self> {
+        let network_passphrase = from_c_string(c.network_passphrase)?;
```
Comment (paulbellamy) [inline cmd/soroban-rpc/lib/preflight/src/lib.rs:43]: Aside: I think we could definitely just calculate the network_id in go and pass that in directly. I don't think preflight every uses the network_passphrase.

File: cmd/soroban-rpc/lib/preflight/src/lib.rs [hunk 5]
```diff
@@ -99,6 +104,7 @@ pub extern "C" fn preflight_invoke_hf_op(
     source_account: *const libc::c_char, // AccountId XDR in base64
     ledger_info: CLedgerInfo,
 ) -> *mut CPreflightResult {
+    println!("gets to top preflight_invoke_hf_op");
```
Comment (paulbellamy) [inline cmd/soroban-rpc/lib/preflight/src/lib.rs:107]: Forgotten?

**Review by paulbellamy — APPROVED** [review]
Now that preflight.rs is independent, it would be great to add some unit tests (which could help show me which bugs you fixed, because right now I've no idea).

**Comment by 2opremio** [comment]
> Now that preflight.rs is independent, it would be great to add some unit tests (which could help show me which bugs you fixed, because right now I've no idea).

Good idea, I will do it in a separate PR though, although the memory leak bugs I fixed were in `lib.rs`

File: cmd/soroban-rpc/lib/preflight/src/fees.rs [hunk 6]
```diff
@@ -361,50 +355,63 @@ pub(crate) fn compute_bump_footprint_exp_transaction_data_and_min_fee(
...
-            );
-        };
+) -> Result<Vec<LedgerEntryRentChange>> {
     let mut rent_changes: Vec<LedgerEntryRentChange> = Vec::new();
-    for key in footprint.read_write.as_vec() {
-        let unmodified_entry = ledger_storage.get(key, true)?;
+    for key in (&footprint).read_only.as_vec() {
+        let unmodified_entry = ledger_storage
+            .get(key, false)
+            .with_context(|| format!("cannot get ledger entry with key {key:?}"))?;
```
Comment (tsachiherman) [inline cmd/soroban-rpc/lib/preflight/src/fees.rs:367]: The key here ( pun intended ) is that we cannot bump the footprint of a key that we can't find on chain. The error that you've provided isn't incorrect, but doesn't deliver the "concept" that a key is expected to be on chain for it's expiration to be bumped.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 1 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:150]: It actually gets them, it only fails if it cannot borrow.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 2 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:176]: Uhm, it is happy either way. I moved everything to the same line.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 3 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:256]: I think it's marshal, thanks

File: cmd/soroban-rpc/lib/preflight/src/fees.rs [hunk 7]
```diff
@@ -304,51 +307,42 @@ pub(crate) fn compute_bump_footprint_exp_transaction_data_and_min_fee(
...
-        rent_changes.push(rent_change);
-    }
+) -> Result<(SorobanTransactionData, i64)> {
+    let rent_changes = compute_bump_footprint_rent_changes(
+        &footprint,
+        ledger_storage,
+        ledgers_to_expire,
+        current_ledger_seq,
+    )
+    .context("cannot compute rent changes")?;
```
Comment (tsachiherman) [inline cmd/soroban-rpc/lib/preflight/src/fees.rs:316]: Could you modify the error message to uniquely identify the error ? i.e. so that it wouldn't be the same error text for both bump and restore ? i.e. `cannot compute bump rent changes`

File: cmd/soroban-rpc/lib/preflight/src/lib.rs (hunk 5 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/lib.rs:107]: Ouch, thanks.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 7 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:316]: Sure, although the operation passed will be different

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 6 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:367]: how would you rephrase it?

File: cmd/soroban-rpc/lib/preflight/src/fees.rs [hunk 8]
```diff
@@ -436,3 +445,41 @@ pub(crate) fn compute_restore_footprint_transaction_data_and_min_fee(
...
+    footprint: &LedgerFootprint,
+    ledger_storage: &LedgerStorage,
+    min_persistent_entry_expiration: u32,
+    current_ledger_seq: u32,
+) -> Result<Vec<LedgerEntryRentChange>> {
+    let mut rent_changes: Vec<LedgerEntryRentChange> = Vec::new();
+    for key in footprint.read_write.as_vec() {
+        let unmodified_entry = ledger_storage
+            .get(key, true)
+            .with_context(|| format!("cannot get ledger entry with key {key:?}"))?;
```
Comment (tsachiherman) [inline cmd/soroban-rpc/lib/preflight/src/fees.rs:457]: could you rephrase that in a more "actionable" tone ? 
i.e. `The ledger key specified in the footprint read-write section {key:?} cannot be found on chain.`

File: cmd/soroban-rpc/lib/preflight/src/fees.rs [hunk 9]
```diff
@@ -436,3 +445,41 @@ pub(crate) fn compute_restore_footprint_transaction_data_and_min_fee(
...
+        let size = (key.to_xdr()?.len() + unmodified_entry.to_xdr()?.len()) as u32;
+        let expirable_entry: Box<dyn ExpirableLedgerEntry> =
+            (&unmodified_entry).try_into().map_err(|e: String| {
+                Error::msg(e.clone()).context("incorrect ledger entry type in footprint")
```
Comment (tsachiherman) [inline cmd/soroban-rpc/lib/preflight/src/fees.rs:461]: could you change this to be more specific ? 
`The ledger key specified in the footprint read-write section was an expirable ledger entry; these entry types cannot be restored directly`

Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:461]: The parent error will tell you that (i.e. `e.clone()`). See `state_expiration.rs`

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    let ConfigSettingEntry::ContractComputeV0(compute) =
+        ledger_storage.get_configuration_setting(ConfigSettingId::ContractComputeV0)?
+        else {
+            bail!("get_budget_from_network_config_params(): unexpected config setting entry for ComputeV0 key");
+        };
+
+    let ConfigSettingEntry::ContractCostParamsCpuInstructions(cost_params_cpu) = ledger_storage
+        .get_configuration_setting(ConfigSettingId::ContractCostParamsCpuInstructions)?
+        else {
+            bail!("get_budget_from_network_config_params(): unexpected config setting entry for ComputeV0 key");
```
Comment (tsachiherman) [inline cmd/soroban-rpc/lib/preflight/src/preflight.rs:190]: ComputeV0 -> ContractCostParamsCpuInstructions

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+        };
+
+    let ConfigSettingEntry::ContractCostParamsMemoryBytes(cost_params_memory) =
+        ledger_storage.get_configuration_setting(ConfigSettingId::ContractCostParamsMemoryBytes)?
+        else {
+            bail!("get_budget_from_network_config_params(): unexpected config setting entry for ComputeV0 key");
```
Comment (tsachiherman) [inline cmd/soroban-rpc/lib/preflight/src/preflight.rs:196]: ComputeV0 -> ContractCostParamsMemoryBytes

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 8 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:457]: I find that unnecessarily verbose

**Review by tsachiherman — COMMENTED** [review]
Few comments; overall looks great.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 1 above)
Comment (paulbellamy) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:150]: Sure, that's fine. I mean that this text we're returning to the user could be misinterpreted.

Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:150]: In reality it's something which should be considered an internal error though.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 8 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:457]: I slightly rephrased it.

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 6 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:366]: OK, I slightly rephrased it.

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+    bucket_list_size: u64,
+    current_ledger_seq: u32,
+) -> Result<PreflightResult> {
+    let (transaction_data, min_fee) = fees::compute_restore_footprint_transaction_data_and_min_fee(
+        footprint,
+        ledger_storage,
+        bucket_list_size,
+        current_ledger_seq,
+    )?;
+    Ok(PreflightResult {
```
Comment (willemneal) [inline cmd/soroban-rpc/lib/preflight/src/preflight.rs:276]: Now a huge issue, but since this is repeated above you could also create a Default implementation and then you could do

```rust
Ok(PreflightResult {
        transaction_data,
        min_fee,
        ..Default::default()
    })
```

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+            }),
+            root_invocation: payload.invocation.clone(),
+        },
+        (None, None) => SorobanAuthorizationEntry {
+            credentials: SorobanCredentials::SourceAccount,
+            root_invocation: payload.invocation.clone(),
+        },
+        // the address and the nonce can't be present independently
+        (a,n) =>
+            panic!("recorded_auth_payload_to_xdr: address and nonce present independently (address: {:?}, nonce: {:?})", a, n),
```
Comment (willemneal) [inline cmd/soroban-rpc/lib/preflight/src/preflight.rs:164]: Why panic when you added `anyhow`?

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+    }
+}
+
+fn host_events_to_diagnostic_events(events: &Events) -> Vec<DiagnosticEvent> {
+    let mut res: Vec<DiagnosticEvent> = Vec::new();
```
Comment (willemneal) [inline cmd/soroban-rpc/lib/preflight/src/preflight.rs:169]: Since we know the length of `events.0` you can use [`Vec::with_capacity`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity), which will prevent extra allocations when pushing.

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs (hunk 12 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/preflight.rs:276]: That’s coming in a separate PR, thanks!!

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs (hunk 14 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/preflight.rs:169]: Good tip, thanks

File: cmd/soroban-rpc/lib/preflight/src/preflight.rs (hunk 13 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/preflight.rs:164]: Fair point

File: cmd/soroban-rpc/lib/preflight/src/fees.rs (hunk 1 above)
Comment (paulbellamy) [reply cmd/soroban-rpc/lib/preflight/src/fees.rs:150]: ok then

**Comment by 2opremio** [comment]
/
